builds/jenkins-build-sls-app/Jenkinsfile

```diff
@@ -0,0 +1,1062 @@
+  if(!providerElem['domain']) providerElem['domain'] = ""
+  def stackTagsElem = providerElem['stackTags']
+  if(!stackTagsElem) {
+    stackTagsElem = [:]
+    providerElem['stackTags'] = stackTagsElem
+  }
+  if(!stackTagsElem['application']) stackTagsElem['application'] = configLoader.INSTANCE_PREFIX
+  if(!stackTagsElem['service']) stackTagsElem['service']="${configLoader.INSTANCE_PREFIX}-${config['domain']}-${config['service']}".toString()
+  if(!stackTagsElem['domain']) stackTagsElem['domain'] = config['domain']
+  if(!stackTagsElem['owner']) stackTagsElem['owner'] = config['created_by']
+  if(!stackTagsElem['environment']) stackTagsElem['environment'] = env
+
+  def customElem = doc['custom']
+  if(!customElem) {
+    customElem = [:]
+    doc['custom'] = customElem
+  }
+  if(!customElem['myStage']) customElem['myStage'] = env
+
+  def functionsElem = doc['functions']
+  if(!functionsElem) {
+    functionsElem = ['handler':[:]]
+    doc['functions'] = functionsElem
+    fillFunctionElem(functionsElem['handler'], 0, config, env)
+  } else {
+    // Traversing through the list of functions and calling fillFunctionElem with index to help with the name creation
+    functionsElem.collect{key, val -> val} // The logical name of function (key) is no use now so we throw it away and mapping values only
+                 .withIndex() // Adding the index to each value rendered. It helps with name creation if such has not been provided
+                 .collect{z -> fillFunctionElem(z[0], z[1], config, env)} // Applying the function for each [val, idx] pair generated
+  }
+
+// Supplying the Outputs element to render all arns for all 'resources/Resources that got listed'
+  def smallResourcesElem = doc['resources']
+  if(smallResourcesElem) {
+    def bigResourcesElem = smallResourcesElem['Resources']
+    if(bigResourcesElem) {
+      def bigOutputsElem = smallResourcesElem['Outputs']
+      if(!bigOutputsElem) {
+        bigOutputsElem = [:]
+        smallResourcesElem['Outputs'] = bigOutputsElem
+      }
+      def resourceKeys = bigResourcesElem.collect{key, val -> key}
+
+      echo "resourceKeys => $resourceKeys"
+      /* Forming a record that extracts the arn from resulting item for each of resource key extracted from resources:
+      UsersTableArn:\n
+        Value:\n
+          \"Fn::GetAtt\": [ usersTable, Arn ]\n
+      */
+      resourceKeys.collect{name ->
+          bigOutputsElem[name+'Arn']=["Value":["Fn::GetAtt":[name, "Arn"]]]
+      }
+    }
+  }
+
+  return doc
+
+}
+
+def servelessDeploy(env, envBucketKey){
+  try {
+    sh "serverless deploy --stage ${env} -v --bucket ${configLoader.AWS.S3[envBucketKey]} > output.log"
+    def outputLog = readFile('output.log').trim()
+    echo "serverless deployment log $outputLog"
+    return outputLog
+  } catch (ex) {
+    echo "Serverless deployment failed due to $ex"
+    def outputLog = readFile('output.log').trim()
+    echo "serverless deployment log $outputLog"
+    throw ex
```

Comment:
Can't throw an exception here. You can refer to other build packs for reference.

Reply:
Well, the exception being thrown is currently intercepted inside the calling method and it is being handled nicely sending all the event about failures to its destinations  
```groovy
        try {
.................................
          echoServerlessFile()
          def deployOutput = servelessDeploy(environment_logical_id, envBucketKey)
.................................
          send_status_email(config, 'COMPLETED', svc_response)

        } catch (ex) {
          send_status_email(config, 'FAILED', '')
          events.sendFailureEvent('UPDATE_ENVIRONMENT', ex.getMessage(), environmentDeploymentMetadata.generateEnvironmentMap("deployment_failed", environment_logical_id, null), environment_logical_id)
          events.sendFailureEvent('UPDATE_DEPLOYMENT', ex.getMessage(), environmentDeploymentMetadata.generateDeploymentMap("failed", environment_logical_id, gitCommitHash), environment_logical_id)
          events.sendFailureEvent('DEPLOY_TO_AWS', ex.getMessage(), context_map, environment_logical_id)
          error ex.getMessage()
        }
```

I would be somewhat reluctant to follow the currently established build-pack pattern where the error is effectively being swallowed and all traces of the error are being lost except the 'echo' statement inside the servelessDeploy(...) function. Please take a look at the entirety of the code inside jenkins-build-pack-lambda/Jenkinsfile
```groovy
        try {
          // initialize aws credentials
          def randomString = utilModule.generateRequestId();
          def credsId = "jazz-${randomString}";
........
          writeServerlessFile(config, environment_logical_id)

          def envBucketKey = "${env_key}${configLoader.JAZZ.S3_BUCKET_NAME_SUFFIX}"
          echoServerlessFile()
          def deployOutput = servelessDeploy(environment_logical_id, envBucketKey)

          if (deployOutput != 'success') {
            if (config['event_source_s3']) {
              handleS3BucketError(deployOutput, config['event_source_s3'], environment_logical_id, envBucketKey, isEventSchdld)
            } else if (config['event_source_sqs']) {
              handleSqsError(deployOutput, config, environment_logical_id, envBucketKey)
            } else if (config['event_source_kinesis']) {
              handleKinesisStreamError(deployOutput, config['event_source_kinesis'], environment_logical_id, envBucketKey)
            } else if (config['event_source_dynamodb']) {
              handleDynamoDbStreamError(deployOutput, config, environment_logical_id, envBucketKey)
            } else {
              echo "Exception occured while serverless deployment to ${environment_logical_id} environment : $deployOutput"
              error "Exception occured while serverless deployment to ${environment_logical_id} environment"
            }
          } else {
...................
          }
          def function_arn = getLambdaARN(stackName, credsId);
...................
          send_status_email(config, 'COMPLETED', svc_response)

        } catch (ex) {
          send_status_email(config, 'FAILED', '')
          events.sendFailureEvent('UPDATE_ENVIRONMENT', ex.getMessage(), environmentDeploymentMetadata.generateEnvironmentMap("deployment_failed", environment_logical_id, null), environment_logical_id)
          events.sendFailureEvent('UPDATE_DEPLOYMENT', ex.getMessage(), environmentDeploymentMetadata.generateDeploymentMap("failed", environment_logical_id, gitCommitHash), environment_logical_id)
          events.sendFailureEvent('DEPLOY_TO_AWS', ex.getMessage(), context_map, environment_logical_id)
          error ex.getMessage()
        }

```
```groovy
def servelessDeploy(env, envBucketKey){
  try {
    sh "serverless deploy --stage ${env} -v --bucket ${configLoader.AWS.S3[envBucketKey]} > output.log"
    return "success"
  } catch (ex) {
    echo "Serverless deployment failed due to $ex"
  }

  def outputLog = readFile('output.log').trim()
  echo "serverless deployment log $outputLog"
  return outputLog
}
```
```groovy
def handleS3BucketError(deployOutput, s3BucketName, environment_logical_id, envBucketKey, isEventSchdld) {
  def s3BucketErrString = "CloudFormation - CREATE_FAILED - AWS::S3::Bucket - "
  s3BucketName = lambdaEvents.getEventResourceNamePerEnvironment(s3BucketName, environment_logical_id, "-")
  if (deployOutput.contains(s3BucketErrString)) {
    if (lambdaEvents.checkS3BucketExists(s3BucketName)) {
      isLambdaUpdateRequired = true
      lambdaEvents.removeS3EventsFromServerless(isEventSchdld)
      deleteAndRedeployService(environment_logical_id, envBucketKey)
    } else {
      error "Error occured while accessing the s3 bucket"
    }
  } else {
    echo "Exception occured while serverless deployment to ${environment_logical_id} environment : $deployOutput"
    error "Exception occured while serverless deployment to ${environment_logical_id} environment"
  }
}
```
Please take into account the excerpt from the following: https://jenkins.io/doc/pipeline/steps/workflow-basic-steps/
"Signals an error. Useful if you want to conditionally abort some part of your program. You can also just throw new Exception(), but this step will avoid printing a stack trace." 
Thus what happens in the current jenkins-build-pack-lambda/Jenkinsfile reads as following: 
1. The method calls a servelessDeploy(...) function
2. A servelessDeploy(...) function swallows the exception received from underlying layer and returns back some gibberish to indicate failure
3. The method recognizes that the returned value is not the expected 'success' value and calls another function that is handleS3BucketError
4. The  handleS3BucketError function does its business and throws the exception with the most generic message that is 'Exception occured while serverless deployment to ${environment_logical_id} environment'. Please note that 
```groovy
error "Exception occured while serverless deployment to ${environment_logical_id} environment"
``` does exactly that
5. The method catches the exception and handles its nicely before re-throwing the exception with the generic method.
As you can see we still throw exceptions even though camouflaged with 'error' statement and we lose the context in the middle that I wanted to preserve in the solution I have implemented

Reply:
:) Agreed, one of the items in technical backlog - refactor the groovy files to do a better job at exception handling & not swallowing actual reasons for errors.